src/nbl/video/CVulkanPhysicalDevice.cpp

```diff
@@ -895,6 +897,43 @@ std::unique_ptr<CVulkanPhysicalDevice> CVulkanPhysicalDevice::create(core::smart
         properties.limits.shaderFloat16 = vulkan12Features.shaderFloat16;
         if (!vulkan12Features.shaderInt8)
             return nullptr;
+
+        if (!properties.limits.shaderSignedZeroInfNanPreserveFloat32)
+            return nullptr;
+        if (!properties.limits.shaderDenormPreserveFloat32)
+            return nullptr;
+        if (!properties.limits.shaderDenormFlushToZeroFloat32)
+            return nullptr;
+        if (!properties.limits.shaderRoundingModeRTEFloat32)
+            return nullptr;
+        if (!properties.limits.shaderRoundingModeRTZFloat32)
+            return nullptr;
+
+        if (vulkan12Features.shaderFloat16) {
+            if (!properties.limits.shaderSignedZeroInfNanPreserveFloat16)
+                return nullptr;
+            if (!properties.limits.shaderDenormPreserveFloat16)
+                return nullptr;
+            if (!properties.limits.shaderDenormFlushToZeroFloat16)
+                return nullptr;
+            if (!properties.limits.shaderRoundingModeRTEFloat16)
+                return nullptr;
+            if (!properties.limits.shaderRoundingModeRTZFloat16)
+                return nullptr;
+        }
+
+        if (deviceFeatures.features.shaderFloat64) {
+            if (!properties.limits.shaderSignedZeroInfNanPreserveFloat64)
+                return nullptr;
+            if (!properties.limits.shaderDenormPreserveFloat64)
+                return nullptr;
+            if (!properties.limits.shaderDenormFlushToZeroFloat64)
+                return nullptr;
+            if (!properties.limits.shaderRoundingModeRTEFloat64)
+                return nullptr;
+            if (!properties.limits.shaderRoundingModeRTZFloat64)
+                return nullptr;
+        }
```

Comment:
if you don't expose them, then you need to move up the Vk Structs and check from the VkStructs, not our own limits struct

Reply:
so query vk features in `Get physical device's limits/properties` scope?

Reply:
for a limit thats required pending a feature, don't do `Vulkan Limits -> Nabla Limit` to then check `Nabla Feature Predicate && not Nabla  Limit`  do `Nabla Feature && ! Vulkan Limit`

because keeping it as is, would require you to expose a limit to Nabla which has no business being exposed (i.e. we can assume the limit is fullfilled if we have the feature).

Example in plain english, if you have 16 bit floats, you can acount on having float controls for float16 type ops. If you don't then whether you would have had controls is irrelevant because you can't use 16bit floats.

Reply:
by "Nabla feature", you mean features defined in `device_features.json`? `shaderFloat16/64` are not defined there but in `device_limit.json` so they are a limit

Reply:
sorry I forgot `shaderFloat` and `shaderInt` are nabla limits, but still you'd want to check
```cpp
if (properties.limits.shaderFloat16 && !vkShaderFloatControlProperties.shaderRoundingModeRTZFloat16)
   return nullptr;
```

so that `properties.limits.shaderRoundingModeRTZFloat16` need not exist

Reply:
`vkShaderFloatControlProperties` isn't a thing and we can't access `vulkan12Properties` after limits are initialized if that's what you meant. because it's in a different scope

Reply:
we also can't *initialize `properties.limits` first* because `shaderFloat16/64` are features not properties

Reply:
> `vkShaderFloatControlProperties` isn't a thing and we can't access `vulkan12Properties` after limits are initialized if that's what you meant. because it's in a different scope

then hoist the variable above the scope...

Reply:
next time suggest solutions while listing problems